File: JAVA/ElgamalECCEncryption.java
package JavaStarterCode;

import java.util.Scanner;
import java.util.Vector;

public class ElgamalECCEncryption {

    /*
    generateReferencePoint: Generate Base point for elliptic curve
    Input:-
        curvePoints: List of points on the Elliptic Curve
    Output:-
        Return Base point or Generator
     */
    private static Point generateReferencePoint(Vector<Point> curvePoints, Helper h)
    {
        Point G = h.randomPoint(curvePoints);
        if(G.getX()==0 && G.getY()==0)
            G = generateReferencePoint(curvePoints, h);
        return G;
    }

    public static void main(String[] args){
        //Elliptic Curve Parameters
        Long a,b,m;
        a = (long)17;
        b = (long)50;
        m = (long)191;


        EllipticCurve ec = new EllipticCurve(a, b, m);
        Vector<Point> curvePoints = ec.getCurvePoints();

        Helper h = new Helper();
        h.generateCharacterPointMappings(curvePoints);

        //Base Point/Generator of Elliptic Curve
        Point G = ElgamalECCEncryption.generateReferencePoint(curvePoints, h);
        Elgamal el = new Elgamal(G, (long)curvePoints.size(), h, ec);

        //Sender and Receiver's Public, Private keys
        Key sender = el.generateKey(curvePoints.size());
        Key receiver = el.generateKey(curvePoints.size());


        String message;
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter message: ");
        message = sc.nextLine();
        System.out.println("Original message: "+message);

        String encryptedMessage = el.Encrypt(message, receiver.getPublicKey(), sender.getPrivateKey());
        System.out.println("Encrypted message: "+encryptedMessage);

        String decryptedMessage = el.Decrypt(encryptedMessage, sender.getPublicKey(), receiver.getPrivateKey());
        System.out.println("Decrypted message: "+decryptedMessage);
    }
}

File: Python/ElgamalECCEncryption.py
from EllipticCurve import EllipticCurve
from Elgamal import Elgamal
from Helper import Helper
from Key import Key
from Point import Point

'''
generateReferencePoint: Generate Base point for elliptic curve
	Input:-
		curvePoints: List of points on the Elliptic Curve
	Output:-
		Return Base point or Generator
'''
def generateReferencePoint(curvePoints, h):
	G = h.randomPoint(curvePoints)
	if G.getX() == 0 and G.getY() == 0:
		G = generateReferencePoint(curvePoints, h)
	return G

def main():
	# Elliptic Curve Parameters
	a = 17
	b = 50
	m = 191

	ec = EllipticCurve(a, b, m)
	curvePoints = ec.getCurvePoints()

	h = Helper()
	h.generateCharacterPointMappings(curvePoints)

	# Base Point/Generator of Elliptic Curve
	G = generateReferencePoint(curvePoints, h)
	el = Elgamal(G, len(curvePoints), h, ec)

	# Sender and Receiver's Public, Private keys
	sender = el.generateKey(len(curvePoints))
	receiver = el.generateKey(len(curvePoints))

	message = input("Enter message: ")
	print("Original message:", message)

	encryptedMessage = el.Encrypt(message, receiver.getPublicKey())
	print("Encrypted message:", encryptedMessage)

	decryptedMessage = el.Decrypt(encryptedMessage, receiver.getPrivateKey())
	print("Decrypted message:", decryptedMessage)

if __name__ == "__main__":
	main()
